docs/DesignDocs/ExtendedEnclaveInitializationData.md

```diff
@@ -0,0 +1,154 @@
+Extended Enclave Initialization Data
+=====
+
+We present an extension of Intel SGX attestation in OpenEnclave that enables the attestation of enclave initialization parameters.
+Our extension uses MRENCLAVE as a platform configuration register to measure additional data besides the signed enclave image, such as instance-specific enclave configuration and the settings in the current enclave signing configuration.
+
+We introduce a new OE quote type for SGX quotes with extended enclave initialization data (EEID).
+Recipients of such quotes must be able to recover and verify the original enclave identity information (including the original MRENCLAVE and MRSIGNER of the base image), based on the knowledge of the initialization data, which is transmitted alongside the quote as a new collateral type.
+
+
+Motivations
+----------
+
+With current SGX quotes, applications can attest their configuration as part of the `reportData` argument of `oe_get_report`.
+This mechanism works robustly when the application code is entirely and statically known to the validator of the quote.
+
+Things get more complicated if an enclave executes user code that is not part of the signed enclave image. For instance, consider an enclave containing a JavaScript interpreter that executes user scripts from the host and has access to the `oe_get_report` API. It is impossible to know which script has been executed by such an enclave based on a standard quote, even if the hash of the script is included in the `reportData`, because a malicious script can obtain a valid quote with `oe_get_report` pretending to be a honest script. Similarly, if an enclave loads and execute arbitrary assembly code from the host, this assembly code can use the `EGETQUOTE` instruction to create valid reports - hence, it is impossible to use attestation to tell which assembly code has been loaded by the enclave.
```

Comment:
This paragraph again seems to convolute evidence format and API.
The agnostic API supports (or needs to support) layered attestation, such as DICE cert chains, for TrustZone.
Once you have layered attestation, then I believe you already have a solution to the interpreter issue being discussed here.  See https://tools.ietf.org/html/draft-ietf-rats-architecture-02#section-4.2 for the IETF summary.

Reply:
To extrapolate Dave's point a bit, I'm inferring that this design needs a layered _measurement_ framework to support the kind of layered attestation; the desire to bind the attestation promises to SGX hardware root of trust implies injecting custom measurements into the only avenue that SGX1 provides for this, which is at sign/load time.

Independent of the EEID proposal, it's worth considering whether the loader can be replaced in a customizable way to permit this, and also to tie that in to a contract with attester plugin registration (i.e. forms of attestation that are predicated on a specific load format must understand whether they can be appropriately supported and fail early if not.)

Reply:
A customizable loader is a great idea - I think that's the only bit that can't go into the current plugin infrastructure. It sounds like major surgery to me though. Are there other extensions that would benefit from that?

Reply:
@CodeMonkeyLeet I suppose it depends on how its implemented. A lot of features on OE take hard dependencies on where/how things are laid out in memory. Did you have a thought about what exactly the developer would be able to customize?

i.e. are we talking wholesale replacement of the loader, or are we talking passed parameters to be able to inform the loader about abitrary things the application would like loaded/measured?

Reply:
@jhand2 Yes, depending on the TEE implementation, this could be a wholesale replacement of the loader (c.f. OP-TEE). We will likely set up a SIG to discuss in greater detail where to take this moving forwards.